contracts/contracts/Tally.sol

```diff
@@ -0,0 +1,248 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+import {AccQueue} from "./trees/AccQueue.sol";
+import {IMACI} from "./IMACI.sol";
+import {Hasher} from "./crypto/Hasher.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {Poll} from "./Poll.sol";
+import {MessageProcessor} from "./MessageProcessor.sol";
+import {SnarkCommon} from "./crypto/SnarkCommon.sol";
+import {Verifier} from "./crypto/Verifier.sol";
+import {VkRegistry} from "./VkRegistry.sol";
+
+
+contract Tally is
+    Ownable,
+    SnarkCommon,
+    Hasher
+{
+    // Error codes
+    string constant ERROR_VOTING_PERIOD_NOT_PASSED = "TallyE01";
+    string constant ERROR_PROCESSING_NOT_COMPLETE = "TallyE02";
+    string constant ERROR_INVALID_TALLY_VOTES_PROOF = "TallyE03";
+    string constant ERROR_ALL_BALLOTS_TALLIED = "TallyE04";
+
+    uint8 private constant LEAVES_PER_NODE = 5;
+
+    // The commitment to the tally results. Its initial value is 0, but after
+    // the tally of each batch is proven on-chain via a zk-SNARK, it should be
+    // updated to:
+    //
+    // hash3(
+    //   hashLeftRight(merkle root of current results, salt0)
+    //   hashLeftRight(number of spent voice credits, salt1),
+    //   hashLeftRight(merkle root of the no. of spent voice credits per vote option, salt2)
+    // )
+    //
+    // Where each salt is unique and the merkle roots are of arrays of leaves
+    // TREE_ARITY ** voteOptionTreeDepth long.
+    uint256 public tallyCommitment;
+
+    uint256 public tallyBatchNum;
+
+    // The final commitment to the state and ballot roots
+    uint256 public sbCommitment;
+
+    Verifier public verifier;
+
+    constructor(Verifier _verifier) {
+        verifier = _verifier;
+    }
+
+    modifier votingPeriodOver(Poll _poll) {
```

Comment:
let's move this to a private function, and add where appropriate in the functions. will be sligthly cheaper

Reply:
do you mean by this:
```
private modifier votingPeriodOver(Poll _poll) {
}
```

Reply:
sorry what I meant is rather than using a modifier we could use a private/internal function which does the checks of the modifier:

```js
function _votingPeriodOver(Poll _poll) private view {
   if (x != y) revert SomeError();
}
```

up to you if you want to go down this route and change all modifiers to private functions. Contract size will be smaller and cheaper to deploy, might result in a small const increase on the user side though.

Reply:
yeah, use the internal function instead based on our discussion offline.